fix(core): CloudFormation dynamic references can't be assigned to num…

madeline-k opened the pull request:
…ber properties

We can't validate the type of dynamic references at synth time, since they will not be resolved until deploy time. This commit removes the type validation check for strings in the format: '{{resolve:...}'.

fixes #14824


----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

madeline-k commented:
Looking for guidance on the best way to test this change. I validated manually with a cdk app the repros the customer issue in https://github.com/aws/aws-cdk/issues/14824 and https://github.com/aws/aws-cdk/issues/14876.  But I am not sure whether it makes more sense to test this with unit tests or integ tests, and in which module(s) those tests should be added.

skinny85 commented:
@madeline-k I would add unit tests for this in the SecretsManager module, verifying that I can pass `Token.asNumber(mySecret. secretValue)` to a property typed as `number` (I would probably use some L1 property whose type is `number` in the test).

madeline-k commented:
@skinny85 Thanks for the guidance! I added a unit test as you described, but that unit test does not actually fail prior to the change in this CR. The error occurs during synth time, so I also added an integ test which does fail prior to the change here. Let me know what you think!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 560eefff0438bdb01f522234b65309aa6bc91659
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=417758bf-ba59-49ac-94e1-17d61b2990b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts:
```diff
@@ -687,6 +687,17 @@ test('fromSecretCompleteArn - grants', () => {
+    runtime: lambda.Runtime.NODEJS,
+    memorySize: cdk.Token.asNumber(secret.secretValueFromJson('LambdaFunctionMemorySize')),
+  });
+});
```
You need to synthesize the Stack for the error to be seen 🙂. Adding any assertion, like `expect(stack).to(haveResourceLike('AWS::Lambda::Function')`, or using `SynthUtils.synthesize()`, will do the trick 🙂.

madeline-k replied:
Oh makes sense, of course! Will update :)

skinny85 reviewed (changes requested):
> @skinny85 Thanks for the guidance! I added a unit test as you described, but that unit test does not actually fail prior to the change in this CR. The error occurs during synth time, so I also added an integ test which does fail prior to the change here. Let me know what you think!

Yes, because your unit test currently doesn't synthesize the Stack 😉. I explained it in my comment below.

I would get rid of the integ test, and rely on the unit test here.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: c4570c4d4f92d5d009f1d6bca477fa37a8f68fe5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2553b13c-2845-4f25-adfd-1ad4ac6c7e4e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts:
```diff
@@ -687,6 +687,19 @@ test('fromSecretCompleteArn - grants', () => {
+    memorySize: cdk.Token.asNumber(secret.secretValueFromJson('LambdaFunctionMemorySize')),
+  });
+
+  SynthUtils.synthesize(stack);
```
Let's write an `expect` assertion here - that checks `memorySize` is `'{{::resolve:secretsmanager...`.

madeline-k replied:
Updated!

skinny85 reviewed (changes requested):
Production code looks great - one small comment on the tests!

skinny85 reviewed (approved):
Beatiful!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 09a1b11763e427a554fe91213bac747e4cb44431
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=44a37a49-8862-4154-ac9d-09d201b220f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4a41afe59abd3960017a13e47bdcbf7bdd5bf557
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9e558c0a-e8ac-44d8-aaaa-2f008413bb86%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
